lib/oxsdatatypes/src/parser.rs

```diff
@@ -125,412 +83,525 @@ impl From<DateTimeError> for XsdParseError {
-            opt(timezone_frag),
-        )),
-        |(year, _, month, _, day, timezone)| Date::new(year, month, day, timezone),
-    )(input)
+// [20]   gYearLexicalRep ::= yearFrag timezoneFrag?
+fn g_year_lexical_rep(input: &str) -> Result<(GYear, &str), XsdParseError> {
+    let (year, input) = year_frag(input)?;
+    let (timezone_offset, input) = optional_end(input, timezone_frag)?;
+    Ok((GYear::new(year, timezone_offset)?, input))
 }
 
-// [19]   gYearMonthLexicalRep ::= yearFrag '-' monthFrag timezoneFrag?
-pub fn g_year_month_lexical_rep(input: &str) -> XsdResult<'_, GYearMonth> {
-    map_res(
-        tuple((year_frag, char('-'), month_frag, opt(timezone_frag))),
-        |(year, _, month, timezone)| GYearMonth::new(year, month, timezone),
-    )(input)
+pub fn parse_g_year(input: &str) -> Result<GYear, XsdParseError> {
+    safe_parse(input, g_year_lexical_rep)
 }
 
-// [20]   gYearLexicalRep ::= yearFrag timezoneFrag?
-pub fn g_year_lexical_rep(input: &str) -> XsdResult<'_, GYear> {
-    map_res(
-        tuple((year_frag, opt(timezone_frag))),
-        |(year, timezone)| GYear::new(year, timezone),
-    )(input)
+// [21]   gMonthDayLexicalRep ::= '--' monthFrag '-' dayFrag timezoneFrag?   Constraint:  Day-of-month Representations
+fn g_month_day_lexical_rep(input: &str) -> Result<(GMonthDay, &str), XsdParseError> {
+    let input = expect_char(input, '-', "gMonthDay values must start with '--'")?;
+    let input = expect_char(input, '-', "gMonthDay values must start with '--'")?;
+    let (month, input) = month_frag(input)?;
+    let input = expect_char(input, '-', "The month and day must be separated by '-'")?;
+    let (day, input) = day_frag(input)?;
+    let (timezone_offset, input) = optional_end(input, timezone_frag)?;
+    Ok((GMonthDay::new(month, day, timezone_offset)?, input))
 }
 
-// [21]   gMonthDayLexicalRep ::= '--' monthFrag '-' dayFrag timezoneFrag?   Constraint:  Day-of-month Representations
-pub fn g_month_day_lexical_rep(input: &str) -> XsdResult<'_, GMonthDay> {
-    map_res(
-        tuple((
-            char('-'),
-            char('-'),
-            month_frag,
-            char('-'),
-            day_frag,
-            opt(timezone_frag),
-        )),
-        |(_, _, month, _, day, timezone)| GMonthDay::new(month, day, timezone),
-    )(input)
+pub fn parse_g_month_day(input: &str) -> Result<GMonthDay, XsdParseError> {
+    safe_parse(input, g_month_day_lexical_rep)
 }
 
 // [22]   gDayLexicalRep ::= '---' dayFrag timezoneFrag?
-pub fn g_day_lexical_rep(input: &str) -> XsdResult<'_, GDay> {
-    map_res(
-        tuple((
-            char('-'),
-            char('-'),
-            char('-'),
-            day_frag,
-            opt(timezone_frag),
-        )),
-        |(_, _, _, day, timezone)| GDay::new(day, timezone),
-    )(input)
+fn g_day_lexical_rep(input: &str) -> Result<(GDay, &str), XsdParseError> {
+    let input = expect_char(input, '-', "gDay values must start with '--'")?;
```

Comment:
Should the error message be `"gDay values must start with '---'"` with three dashes?